File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.*;

class Solution {
    public int[] solution(int N, int[] stages) {
        int[] answer = new int[N];

        int[] count = new int[N + 2];

        for (int stage : stages) {
            count[stage]++;
        }

        Map<Integer, Double> map = new HashMap<>();

        int remain = stages.length;

        for (int i = 1; i <= N; i++) {
            if (remain == 0) {
                map.put(i, 0.0);
            } else {
                map.put(i, (double) count[i] / remain);
            }

            remain -= count[i];
        }

        List<Map.Entry<Integer, Double>> list = new ArrayList<>(map.entrySet());

        list.sort((a, b) -> {
            int cmp = Double.compare(b.getValue(), a.getValue());

            if (cmp == 0) {
                return Integer.compare(a.getKey(), b.getKey());
            }

            return cmp;
        });

        for (int i = 0; i < N; i++) {
            answer[i] = list.get(i).getKey();
        }

        return answer;
    }
}

File: 프로그래머스/1/42889. 실패율/실패율.py
from collections import Counter
def solution(N, stages):
    count = Counter(stages)
    answer = []
    total = len(stages)

    for stage in range(1, N + 1):
        if total == 0:
            failure = 0
        else:
            failure = count[stage] / total

        answer.append((stage, failure))
        total -= count[stage]

    answer.sort(key=lambda x: (-x[1], x[0]))

    return [stage for stage, failure in answer]